### PR TITLE
Futhark Formatter

### PR DESCRIPTION
**Pull request by WilliamDue**
This pull request contains the work in progress code formatter for Futhark.

### CODE REVIEW

**Comment by athas** [comment]
If you want submodules, then put them under `Futhark.Fmt`. There is one rule for the `Futhark.CLI` hierarchy, and that is that the modules there corrspond exactly to CLI executables.

**Review by athas — CHANGES_REQUESTED** [review]
The `Futhark.CLI.Fmt.AST` module does not, in fact, define an AST.

Although you should consider the formatting of the following file.

```Haskell
def x = 0x123
```

You will notice that the output is wrong. The solution is to keep around the original source bytestring and use the offsets in the AST location information to retrieve the original syntax for number literals.

**Comment by athas** [comment]
I have pushed a change to `Futhark.CLI.Fmt` that makes it read input from stdin if no file is given.

**Comment by WilliamDue** [comment]
> The `Futhark.CLI.Fmt.AST` module does not, in fact, define an AST.
> 
> Although you should consider the formatting of the following file.
> 
> ```haskell
> def x = 0x123
> ```
> 
> You will notice that the output is wrong. The solution is to keep around the original source bytestring and use the offsets in the AST location information to retrieve the original syntax for number literals.

We should have solved this problem [https://github.com/diku-dk/futhark/pull/2178/commits/80c79c636a6abcfe6445589c67d611af59bf8467](here). Also we will think of a better name for AST.

**Comment by athas** [comment]
Since this is now pretty much functionally correct, I have marked it for review. Fix all the style issues please.

File: tests_fmt/build_exec_test.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
+
+cabal build > /dev/null
```
Comment (athas) [inline tests_fmt/build_exec_test.sh:3]: A test script should not compile the compiler. How would you use this in a CI setting?

File: tests_fmt/test.sh [hunk 2]
```diff
@@ -0,0 +1,107 @@
+#!/bin/sh
```
Comment (athas) [inline tests_fmt/test.sh:1]: There is no need for this to be in a subdirectory. Put it in `tools` along with `testparser.sh`.

File: tests_fmt/build_exec_test.sh (hunk 1 above)
Comment (WilliamDue) [reply tests_fmt/build_exec_test.sh]: We have removed this script.

File: tests_fmt/test.sh (hunk 2 above)
Comment (WilliamDue) [reply tests_fmt/test.sh]: We moved it to `tools` and called it `testformatter.sh`

**Comment by WilliamDue** [comment]
We are decently happy with this version. Currently trailing comments works but it does not work perfectly see `tests/fmt/trailingComments0.fut`. We think we could do a hacky solution but it seems like we would have to maybe rethink some design choices a bit to make it work in a none hacky way.

File: Setup.hs [hunk 3]
```diff
@@ -4,4 +4,5 @@ import Distribution.Simple
 
 main :: IO ()
 main = defaultMainWithHooks myHooks
-  where myHooks = simpleUserHooks
+  where
+    myHooks = simpleUserHooks
```
Comment (athas) [inline Setup.hs:8]: Why did you change this?

File: src/Futhark/Fmt/Monad.hs [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+data LastOutput = Line | Space | Text deriving (Show, Eq, Ord)
+
+-- | A data type to describe the layout the formatter is using currently.
+data Layout = MultiLine | SingleLine deriving (Show, Eq)
+
+-- | The format monad used to keep track of comments and layout. It is a a
+-- combincation of a reader and state monad. The comments and reading from the
+-- input file are the state monads job to deal with. While the reader monad
+-- deals with the propagating the current layout.
+type FmtM a = ReaderT Layout (StateT FmtState Identity) a
```
Comment (athas) [inline src/Futhark/Fmt/Monad.hs:153]: Instead of `StateT FmtState Identity`, just say `State FmtState`.

File: src/Futhark/Fmt/Printer.hs [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+  fmt (SizeExp d loc) = addComments loc $ brackets d
+  fmt (SizeExpAny loc) = addComments loc $ brackets nil
+
+instance Format UncheckedSpec where
+  fmt (TypeAbbrSpec tpsig) = fmt tpsig
+  fmt (TypeSpec l name ps doc loc) =
+    addComments loc $
+      doc <:> text "type" <+> l <:> sub
+    where
+      sub = sepFilter [True, not $ null ps] line [fmtName name, align (sep line ps)]
```
Comment (athas) [inline src/Futhark/Fmt/Printer.hs:431]: This is a very convoluted way of writing
```
sub | null ps = fmtName name
    | otherwise = sep line [fmtName name, align (sep line ps)]

File: src/Futhark/Fmt/Printer.hs [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+  fmt (ValSpec name ps te _ doc loc) =
+    addComments loc $
+      doc
+        <:> text "val"
+        <+> sub
+        <:> text ":"
+        <+> te
+    where
+      sub = sepFilter [True, not $ null ps] line [fmtName name, align (sep line ps)]
```
Comment (athas) [inline src/Futhark/Fmt/Printer.hs:440]: Similar to above. `sepFilter` is a strange function.

**Review by athas — CHANGES_REQUESTED** [review]
The `tests/` directory is for tests that that are run. For testing whether programs are formatted correctly, I suggest you create a `tests_fmt/` directory in the repository root that contains unformatted programs, along with what they are expected to look like after formatting. See `tests_literate/` for a suite of tests that does something similar.

File: Setup.hs (hunk 3 above)
Comment (WilliamDue) [reply Setup.hs:8]: This happened when `tools/run-formatter.sh` was used for formatting.

Comment (WilliamDue) [reply Setup.hs:8]: It was run on all of the the repository at some point.

**Review by athas — COMMENTED** [review]
Please don't reformat files that are not relevant to your work.

**Comment by athas** [comment]
What are all those Zone.Identifier files?

**Comment by WilliamDue** [comment]
In technical terms, we have made whoopsie daisy.

**Comment by WilliamDue** [comment]
> Please don't reformat files that are not relevant to your work.

I think the problem is gone now.

**Comment by athas** [comment]
`tests_fmt/expected` would be more interesting if all the input files were not already formatted.

**Comment by athas** [comment]
Can I merge this now?

**Comment by thereseLyngby** [comment]
Test script in tests_fmt is not done yet, but should be in next hour or so. The test themselves remain "uninteresting", most of them cover issues, where the AST used to represent two different code snippets with the same term (like the pattern thing in loops).

**Comment by athas** [comment]
So it is done now?

**Comment by WilliamDue** [comment]
We have added it to CI now it should hopefully be "done" now. I think we have some edge case problems with correctly place comments dealing with `sep` but they should probably be squashed as they come. One of these are sumtypes, they are not located if they do not take an argument.

**Comment by WilliamDue** [comment]
We think ideally the Fmt type should also be `Located`such that comments can be placed correctly and the location should be based on the expression that is formatted formats. And when concatenated with another Fmt the smallest location that is not NoLoc should be used for the new Fmt.

**Comment by athas** [comment]
`chmod +x tools/testformatter.sh`

**Comment by WilliamDue** [comment]
Actually a side note, there is probably quite alot of cases where `sep` places comments wrongly. This is because we had to change how sep worked because we wanted trailing comments. The thing with `Fmt` having a locaiton is probably not that good actually.
